### PR TITLE
Fota support manifest v3

### PR DESCRIPTION
**Pull request by micgur01**
Change bootloader to support new fota header and manifest v3

### CODE REVIEW

File: fota/fota.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
+// ----------------------------------------------------------------------------
+// Copyright 2018-2019 ARM Ltd.
```
Comment (JanneKiiskila) [inline fota/fota.h:2]: 2018-2020

File: fota/fota_app_ifs.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
+// ----------------------------------------------------------------------------
+// Copyright 2018-2019 ARM Ltd.
```
Comment (JanneKiiskila) [inline fota/fota_app_ifs.c:2]: 2018-2020

File: upgrade.c [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+#if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
+    uint8_t fw_key[FOTA_ENCRYPT_KEY_SIZE];
+    ret = fota_nvm_fw_encryption_key_get(fw_key);
+    if (ret) {
+        pr_info("Candidate not found");
+        return FOTA_STATUS_NOT_FOUND;
+    }
+    memset(fw_key, 0, sizeof(fw_key));
+    force_encrypt = true;
+#endif // #if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
```
Comment (alzix) [inline upgrade.c:215]: i think it is "safer" to initialize `force_encrypt` value at declaration.
```suggestion

#if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
    bool force_encrypt = true;
    uint8_t fw_key[FOTA_ENCRYPT_KEY_SIZE];
    ret = fota_nvm_fw_encryption_key_get(fw_key);
    if (ret) {
        pr_info("Candidate not found");
        return FOTA_STATUS_NOT_FOUND;
    }
    memset(fw_key, 0, sizeof(fw_key));
    force_encrypt = true;
#else
    bool force_encrypt = false;
#endif // #if (MBED_CLOUD_CLIENT_FOTA_ENCRYPTION_SUPPORT == 1)
```

**Comment by alzix** [comment]
@micgur01 i think we should specify the origin of imported FOTA sources.
please consider borrowing the https://github.com/ARMmbed/manifest-tool/blob/master/bsdiff/import.sh script and saving the https://github.com/ARMmbed/manifest-tool/blob/master/bsdiff/import_ref.txt

File: upgrade.c (hunk 3 above)
Comment (micgur01) [reply upgrade.c:215]: I'll change, but remove line 215
why it's safer?

**Comment by JanneKiiskila** [comment]
Any deployment or compatibility issues with older versions?

File: upgrade.c (hunk 3 above)
Comment (alzix) [reply upgrade.c:215]: not sure :) FI mitigations

Comment (alzix) [reply upgrade.c:215]: removed line 215

File: .mbedignore [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+mbed-os/usb/device/USBAudio/*
+mbed-os/drivers/I2CSlave.cpp
+mbed-os/storage/kvstore/source/kvstore_global_api.cpp
+fota/fota_source.c
+fota/fota_manifest_v1.c
+fota/fota_manifest_v3.c
+fota/fota.c
+fota/fota_app_ifs.c
+fota/fota_delta.c
+fota/fota_event_handler.c
```
Comment (SeppoTakalo) [inline .mbedignore]: Why exactly is this file needed?

And why does it list `fota.c` which is coming in this PR?

File: fota/fota.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#endif
+
+
+#include "fota/fota_status.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct endpoint_s endpoint_t;
```
Comment (SeppoTakalo) [inline fota/fota.h]: Why do you typedef something that is coming from the Client headers?
To be consistent I think we should be use the types as they are defined.

**Comment by alzix** [comment]
@JanneKiiskila,
> Any deployment or compatibility issues with older versions?

Yes. This is new bootloader to be deployed on new targets. It is not intended to replace existing bootloader on already deployed targets.
When using new bootloader, FOTA library in client can deliver more features, such as:

File: mbed_app.json [hunk 6]
```diff
@@ -62,47 +36,110 @@
...
+        "platform",
+        "mbed-bootloader",
+        "mbed-trace",
+        "mbedtls",
+        "system-storage",
+        "storage",
+        "storage_tdb_internal",
+        "flashiap-block-device",         
+        "drivers",        
+		"storage_filesystem",
```
Comment (SeppoTakalo) [inline mbed_app.json:49]: Whitespace errors

File: mbed_app.json [hunk 7]
```diff
@@ -62,47 +36,110 @@
...
+        "filesystemstore",
+        "littlefs",
+        "fat_chan",
+        "direct-access-devicekey",
+        "tdbstore",
+        "kv-config",
+        "kv-global-api",
+        "kv-map",
+        "device_key",
+		 "fota"
```
Comment (SeppoTakalo) [inline mbed_app.json:61]: Whitespace problems

File: mbed_lib.json [hunk 8]
```diff
@@ -1,71 +1,53 @@
...
             "help": "The max application size",
             "value": "(MBED_ROM_START + MBED_ROM_SIZE - MBED_CONF_MBED_BOOTLOADER_APPLICATION_START_ADDRESS)",
             "macro_name": "DEFAULT_MAX_APPLICATION_SIZE"
         },
-        "bootloader-size":{
-            "help": "bootloader size",
-            "value": "(32*1024)",
-            "macro_name": "MBED_BOOTLOADER_SIZE"
-        },
-        "startup-delay": {
```
Comment (SeppoTakalo) [inline mbed_lib.json:55]: Why is this functionality removed?

File: mbed_lib.json [hunk 9]
```diff
@@ -1,71 +1,53 @@
...
-        "show-serial-output": {
+        "trace": {
             "help": "Show boot status and progress on serial output. Disable output to save space on headless devices.",
-            "value": 1,
-            "macro_name": "SHOW_SERIAL_OUTPUT"
-        },
-        "show-swo-output": {
-            "help": "Show boot status and progress on SWO output instead of serial.",
-            "value": 0,
-            "macro_name": "SHOW_SWO_OUTPUT"
```
Comment (SeppoTakalo) [inline mbed_lib.json:38]: Why is this removed?

File: mbed_lib.json [hunk 10]
```diff
@@ -1,71 +1,53 @@
...
-        "nonstandard-entrypoint": {
-            "help": "Set to 1 to turn the Mbed Bootloader application into a library that can be ported to other platforms.",
-            "value": null,
-            "macro_name": "MBED_BOOTLOADER_NONSTANDARD_ENTRYPOINT"
-        },
-        "show-serial-output": {
+        "trace": {
             "help": "Show boot status and progress on serial output. Disable output to save space on headless devices.",
-            "value": 1,
-            "macro_name": "SHOW_SERIAL_OUTPUT"
```
Comment (SeppoTakalo) [inline mbed_lib.json:33]: Why is this removed?

File: .mbedignore (hunk 4 above)
Comment (micgur01) [reply .mbedignore]: removed

File: fota/fota.h (hunk 5 above)
Comment (alzix) [reply fota/fota.h]: this is forward declaration - the type is coming from client header files 
@michael - do we even need to import this file?

Comment (SeppoTakalo) [reply fota/fota.h]: Forward declaration makes sense for opaque types.. but I don't see why you changed the name as well.. however, this is minor detail.

Comment (micgur01) [reply fota/fota.h]: it was imported, but no more from latest commit
